subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -0,0 +1,2106 @@
+#include "hfp_ag_internal.h"
+
+#define LOG_LEVEL CONFIG_BT_HFP_AG_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_hfp_ag);
+
+typedef int (*bt_hfp_ag_parse_command_t)(struct bt_hfp_ag *ag, struct net_buf *buf,
+					 enum at_cme *err);
+
+struct bt_hfp_ag_at_cmd_handler {
+	char *cmd;
+	bt_hfp_ag_parse_command_t handler;
+};
+
+static const struct {
+	char *name;
+	char *connector;
+	uint32_t min;
+	uint32_t max;
+} ag_ind[] = {
+	{"service", "-", 0, 1},   /* BT_HFP_AG_SERVICE_IND */
+	{"call", ",", 0, 1},      /* BT_HFP_AG_CALL_IND */
+	{"callsetup", "-", 0, 3}, /* BT_HFP_AG_CALL_SETUP_IND */
+	{"callheld", "-", 0, 2},  /* BT_HFP_AG_CALL_HELD_IND */
+	{"signal", "-", 0, 5},    /* BT_HFP_AG_SINGNAL_IND */
+	{"roam", ",", 0, 1},      /* BT_HFP_AG_ROAM_IND */
+	{"battchg", "-", 0, 5}    /* BT_HFP_AG_BATTERY_IND */
+};
+
+typedef void (*bt_hfp_ag_tx_cb_t)(struct bt_hfp_ag *ag, void *user_data);
+
+struct bt_ag_tx {
+	sys_snode_t node;
+
+	struct bt_hfp_ag *ag;
+	struct net_buf *buf;
+	bt_hfp_ag_tx_cb_t cb;
+	void *user_data;
+};
+
+static void bt_hfp_ag_ringing_cb(struct bt_hfp_ag *ag, void *user_data);
+
+static int hfp_ag_send_data(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *user_data,
+			    const char *format, ...);
+static int hfp_ag_next_step(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *user_data);
+
+static int bt_hfp_ag_brsf_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_bac_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_cind_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_cmer_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_chld_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_bind_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_cmee_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_chup_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_clcc_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_bia_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_ata_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_cops_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_bcc_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_bcs_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_atd_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_bldn_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
+static int bt_hfp_ag_clip_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err);
```

Comment:
Massive amount of forward declarations here. Can you get rid of them by doing some reorganizing within the c-file?

Reply:
Removed all function declarations.